python/nav/smidumps/__init__.py

```diff
@@ -3,3 +3,47 @@
 As dumped by smidump dump using the python format option.
 
 """
+from __future__ import absolute_import
+from os.path import dirname, join, basename, isfile, splitext
+import glob
+import importlib
+
+_submodule_files = (
+    glob.glob(join(dirname(__file__), "*.py")) +
+    glob.glob(join(dirname(__file__), "*.pyc")) +
+    glob.glob(join(dirname(__file__), "*.pyo"))
+)
+
+_submodules = list(set(
+    splitext(basename(f))[0]
+    for f in _submodule_files
+    if isfile(f) and not basename(f).startswith('__init__')
+))
+
+__all__ = _submodule_files = ['get_mib_modules', 'get_mib']
+
+_mib_map = {}
+
+
+def get_mib_modules():
+    """Returns a dict mapping MIB module names to the corresponding Python
+    module containing its smidumped version.
+
+    """
+    if not _mib_map:
+        for modname in _submodules:
+            module = importlib.import_module('.' + modname, 'nav.smidumps')
+            if hasattr(module, 'MIB') and 'moduleName' in module.MIB:
+                _mib_map[module.MIB['moduleName']] = module
```

Comment:
Does this mean we no longer can load one smidump without loading them all? Could that have unforeseen consequences? Some usages (portadmin) might import these from a request context, which might take noticably longer?

Reply:
This interpretation is correct, yes. Since there is currently no way of knowing the mapping between a MIB module name and a python module other than importing them all and checking the contents, that's how it is.

I did not foresee that the modules were used by anything other than ipdevpoll (which loads all of these modules anyway). Regarding this other usage:

1. *snmptrapd*: Not a big deal. snmptrapd is a long running daemon, the overhead of loading all the smidump modules at startup time is negligible.
2. *portadmin*: The module imports should not run on every request to the web server - if they do, something is wrong with the way you are serving NAV.

However: Both snmptrapd and portadmin currently import the Python modules directly, which would presumably bypass the loading of all submodules.

So in short: I do not think has much of an impact as it stands.

Moving forward (not in this PR), if we move to using the `get_mib()` call to load MIBs everywhere, `get_mib()` can easily be refactored later to load the modules in other ways (without loading them all at once). One suggestion is to name the Python modules verbatim after their corresponding MIB modules. That would make them non-importable through regular Python import statements, but easily importable through dynamic code. You can even imagine `get_mib()` using smidump directly in an attempt to load MIB modules that haven't been converted to Python dicts yet.

Reply:
I was thinking that it might be possible to do something to the dumps to make it possible to find the correct one without loading all of them. Either we could rename the modules to match the mib name, or rename them to a python compatible name derived from the mib name, or we could save a mapping in a config file or similar.

As for the web interface it seems all cases where portadmin loads these are done at parse time, so it will only cause somewhat prolonged startup time (which could be annoying in dev context, but not important). It would also somewhat increase the memory footprint of the web server, but also not important probably. And irrelevant as long as the code as you say don't acutally load all modules

Reply:
Since I suggested it above, I would rename the python modules to match the mib name - which is actually possible now that we are basing everything on dynamic imports. I'm not sure it belongs in this PR, though - since it has already become more than large enough :-D